projects/observability/src/pages/custom-dashboards/custom-dashboard-panel-edit/custom-dashboard-panel-edit.component.ts

```diff
@@ -0,0 +1,232 @@
+  ];
+  public constructor(
+    private readonly metadataService: MetadataService,
+    private readonly activatedRoute: ActivatedRoute,
+    protected readonly customDashboardStoreService: CustomDashboardStoreService,
+    private readonly customDashboardService: CustomDashboardService,
+    private readonly navigationService: NavigationService
+  ) {
+    const uniqueRequests$ = this.requestSubject.pipe(distinctUntilChanged(isEqualIgnoreFunctions));
+    this.visualizationDashboard$ = uniqueRequests$.pipe(
+      switchMap(request => this.buildVisualizationDashboard(request))
+    );
+    this.activatedRoute.params.subscribe(params => {
+      this.isNewPanel = params.panel_id === 'new';
+      this.panelId = params.panel_id;
+      this.dashboardId = params.dashboard_id;
+    });
+    this.activatedRoute.queryParams.subscribe(params => {
+      this.dashboardName = params.dashboardName;
+      this.isNewDashboard = params.newDashboard;
+    });
+
+    if (!this.isNewPanel) {
+      const panelData = this.customDashboardStoreService.getPanel(this.dashboardId, this.panelId);
+      if (panelData) {
+        this.state = panelData;
+      } else {
+        // Fallback to listing incase user refereshes on panel edit page. In that case data is not present in the dashboard store since it's in memory.
+        this.navigationService.navigateWithinApp(['/custom-dashboards']);
+
+        return;
+      }
+    } else {
+      this.state = {
+        context: ObservabilityTraceType.Api,
+        resultLimit: 15,
+        series: [],
+        name: 'New Panel',
+        id: '',
+        is_realtime: false,
+        interval: 'AUTO',
+        json: {
+          type: 'cartesian-widget',
+          'selectable-interval': false,
+          'series-from-data': true,
+          'legend-position': LegendPosition.Bottom,
+          'selection-handler': {
+            type: 'cartesian-explorer-selection-handler',
+            'show-context-menu': false
+          }
+        }
+      };
+    }
+    this.currentContext = this.contextItems.find(i => i.value === this.state.context)!;
+    this.setFilters();
+  }
+
+  public onFiltersUpdated(newFilters: Filter[]): void {
+    this.state.filters = [...newFilters];
+  }
+  public setFilters(): void {
+    this.attributes$ = this.metadataService.getFilterAttributes(this.state.context);
+  }
+  public onPanelNameChange(name: string): void {
+    this.state.name = name;
+  }
+  public onContextChange(context: ToggleItem<string>): void {
+    this.state.context = context.value!;
+    this.currentContext = this.contextItems.find(i => i.value === this.state.context)!;
+
+    this.attributes$ = this.metadataService.getFilterAttributes(this.state.context);
+  }
+  public onCheckRealtime(val: boolean): void {
+    this.state.is_realtime = val;
+  }
+  private buildVisualizationDashboard(request: ExploreVisualizationRequest): Observable<ExplorerGeneratedDashboard> {
+    return of({
+      json: cloneDeep(this.state.json),
```

Comment:
Can we remove the deep clone?

Reply:
The internal hda-dashboard components for rendering checks whether we pass a new json object to trigger a the dashboardReady event where we are handling the changes to filters and other fields. Below screenshot is from the hyperdash-angular repo from where the changes are triggered
<img width="571" alt="Screenshot 2022-08-04 at 11 57 32 AM" src="https://user-images.githubusercontent.com/101543426/182778213-cd57ad33-1ed9-4f56-8cfc-893144de6b81.png">

Reply:
Interesting. Please add a comment to the code explaining the reasoning behind deep clone. This would be hard to deduce if not explained.